tracer/tracer.go

```diff
@@ -499,129 +571,172 @@ func loadAllMaps(coll *cebpf.CollectionSpec, ebpfMaps map[string]*cebpf.Map,
-		},
-		{
-			progID: uint32(support.ProgUnwindV8),
-			name:   "unwind_v8",
-			enable: includeTracers.Has(types.V8Tracer),
-		},
-		{
-			progID: uint32(support.ProgUnwindDotnet),
-			name:   "unwind_dotnet",
-			enable: includeTracers.Has(types.DotnetTracer),
-		},
-		{
-			name:             "tracepoint__sched_process_exit",
+	progs := make([]progLoaderHelper, len(tailCallProgs)+2)
+	copy(progs, tailCallProgs)
+	progs = append(progs,
+		progLoaderHelper{
+			name:             "finish_task_switch",
 			noTailCallTarget: true,
 			enable:           true,
 		},
-		{
-			name:             "native_tracer_entry",
+		progLoaderHelper{
+			name:             "tracepoint__sched_switch",
 			noTailCallTarget: true,
 			enable:           true,
 		},
-	} {
+	)
+
+	for _, unwindProg := range progs {
 		if !unwindProg.enable {
 			continue
 		}
 
-		// Load the eBPF program into the kernel. If no error is returned,
-		// the eBPF program can be used/called/triggered from now on.
-		unwinder, err := cebpf.NewProgramWithOptions(coll.Programs[unwindProg.name],
-			programOptions)
-		if err != nil {
-			// These errors tend to have hundreds of lines (or more),
-			// so we print each line individually.
-			if ve, ok := err.(*cebpf.VerifierError); ok {
-				for _, line := range ve.Log {
-					log.Error(line)
-				}
-			} else {
-				scanner := bufio.NewScanner(strings.NewReader(err.Error()))
-				for scanner.Scan() {
-					log.Error(scanner.Text())
-				}
+		progSpec, ok := coll.Programs[unwindProg.name]
+		if !ok {
+			return fmt.Errorf("program %s does not exist", unwindProg.name)
+		}
+
+		// Replace the prog array for the tail calls.
+		insns := progArrayReferences(perfTailCallMapFD, progSpec.Instructions)
+		for _, ins := range insns {
+			if err := progSpec.Instructions[ins].AssociateMap(tailcallMap); err != nil {
+				return fmt.Errorf("failed to rewrite map ptr: %v", err)
 			}
-			return fmt.Errorf("failed to load %s", unwindProg.name)
 		}
 
-		ebpfProgs[unwindProg.name] = unwinder
-		fd := uint32(unwinder.FD())
-		if unwindProg.noTailCallTarget {
-			continue
+		// All the tail call targets are perf event programs. To be able to tail call them
+		// from a kprobe, adjust their specification.
+		if !unwindProg.noTailCallTarget {
+			// Adjust program type
+			progSpec.Type = cebpf.Kprobe
+
+			// Adjust program name for easier debugging
+			progSpec.Name = "kp_" + progSpec.Name
 		}
-		if err := tailcallMap.Update(unsafe.Pointer(&unwindProg.progID), unsafe.Pointer(&fd),
-			cebpf.UpdateAny); err != nil {
-			// Every eBPF program that is loaded within loadUnwinders can be the
-			// destination of a tail call of another eBPF program. If we can not update
-			// the eBPF map that manages these destinations our unwinding will fail.
-			return fmt.Errorf("failed to update tailcall map: %v", err)
+		if err := loadProgram(ebpfProgs, tailcallMap, unwindProg.progID, progSpec,
+			programOptions, unwindProg.noTailCallTarget); err != nil {
+			return err
```

Comment:
While this is clever, and avoids some function declarations. I think this might have some maintenance overhead. Especially if anything in the ebpf binary becomes more function type specific. And this also prevents to have differing probe context usage for perf_event vs. kprobe. Perhaps in future we would like the perf_event probe to also record perf event specific data from the context?

I would prefer a solution to do this in the eBPF C code instead. Just separate is tracepoint to main function that gets called from each ebpf helper entry point separately. This would be future proof without maintenance fear. Even if it likely doubles the eBPF ELF binary size.

Reply:
For the moment, I will keep the proposed solution. Currently there is a high demand to also check in the debug eBPF blobs - which would add binary blobs of 1.4M (amd64) + 1.4M (arm64) to the respository.
 
 With a macro like the following it is possible to generate perf event and kprobe programs from the same source, avoiding code duplication:
```
#define MULTI_USE_FUNC(func_name) \
    SEC("perf_event/"#func_name) \
    int func_name##_perf(struct pt_regs *ctx) { \
        return func_name(ctx); \
    } \
    \
    SEC("kprobe/"#func_name) \
    int func_name##_kprobe(struct pt_regs *ctx) { \
        return func_name(ctx); \
    }
```

If it is decided to also check in debug eBPF blobs, then the checked in binary blobs will be the sum of 1.4M (amd64/perf events) + 1.4M (arm64/perf events) +  1.4M (amd64/kprobe) + 1.4M (arm64/kprobe)  just for the debug binary eBPF blobs.
Without the discussion around also keeping the debug binary eBPF blobs checked into the repository, switching to the marco for code generation sounds like a viable option. But with this discussion, I would like to defer this implementation detail for a moment and wait for the discussion result.

While using a macro to generate these two different kinds of eBPF programs from the same code is possible and avoid relabeling the eBPF programs, the more significant part is updating the tail call. I didn't find a way to rewrite the tail call maps when generating code with the above marco. So dynamic rewriting of the tail call map would stay anyway.

Independent of the way it is done in the end, the relabeling or program generation using a macro only affects the programs that are tail call destinations and are used for unwinding the stack. The entry hooks, no matter whether it is a perf event type of program of kprobe, will stay native eBPF programs and not touched. This allows to do eBPF program specific tasks - only the tail call destinations are generic and not eBPF program type specific. So there is no limitation in fetching and handling eBPF program type specific information.

Reply:
opened https://github.com/open-telemetry/opentelemetry-ebpf-profiler/issues/225 to bring this discussion forward.

Reply:
With https://github.com/open-telemetry/opentelemetry-ebpf-profiler/pull/196/commits/c1724aa71e9b25146fbabb9f6e4c860b83a09daa I have rewritten the approach to generate perf event and kprobe programs at compile time as requested.
With this change the new size of `tracer.ebpf.release.amd64` is 570 KBytes (before 291 KBytes) and for `tracer.ebpf.release.arm64` it becomes 554 KBytes (before 283 KBytes). 


Let me know what you think - @fabled @christos68k

Reply:
friendly ping for feedback @fabled and @christos68k  - or can this conversation be resolved?

Reply:
friendly ping for feedback @fabled and @christos68k - or can this conversation be resolved?

Reply:
friendly ping for feedback @fabled and @christos68k - or can this conversation be resolved?

Reply:
We decided not to commit the debug artifacts in the repository so I think we can proceed with @fabled suggestion (which AFAICT is what you implemented in c1724aa)